query/PRoST-query/src/main/java/translator/Translator.java

```diff
@@ -267,44 +242,61 @@ public Node buildTree(final List<Triple> triples, final List<Var> projectionVars
 				// create nodes
 				if (biggestObjectGroupSize > biggestSubjectGroupSize) {
 					// create and add the rpt or vp node
-					if (biggestObjectGroupSize >= minimumGroupSize) {
-						nodesQueue.add(new IptNode(biggestObjectGroupTriples, prefixes));
-					} else {
-						for (final Triple t : biggestObjectGroupTriples) {
-							final String tableName = Stats.getInstance().findTableName(t.getPredicate().toString());
-							final Node newNode = new VpNode(new TriplePattern(t, prefixes), tableName);
-							nodesQueue.add(newNode);
-						}
-					}
-					removeTriplesFromGroups(biggestObjectGroupTriples, subjectGroups); // remove empty groups
-					objectGroups.remove(biggestObjectGroupIndex); // remove group of created node
+					createRPtVPNode(biggestObjectGroupTriples, nodesQueue);
+					removeTriplesFromGroups(biggestObjectGroupTriples, subjectGroups);
+					objectGroups.remove(biggestObjectGroupIndex);
 				} else {
 					/// create and add the pt or vp node
-					if (biggestSubjectGroupSize >= minimumGroupSize) {
-						nodesQueue.add(new PtNode(biggestSubjectGroupTriples, prefixes));
-					} else {
-						for (final Triple t : biggestSubjectGroupTriples) {
-							final String tableName = Stats.getInstance().findTableName(t.getPredicate().toString());
-							final Node newNode = new VpNode(new TriplePattern(t, prefixes), tableName);
-							nodesQueue.add(newNode);
-						}
-					}
-					removeTriplesFromGroups(biggestSubjectGroupTriples, objectGroups); // remove empty groups
-					subjectGroups.remove(biggestSubjectGroupIndex); // remove group of created node
+					createPtVPNode(biggestSubjectGroupTriples, nodesQueue);
+					removeTriplesFromGroups(biggestSubjectGroupTriples, objectGroups);
+					subjectGroups.remove(biggestSubjectGroupIndex);
 				}
 			}
 		} else {
 			// VP only
 			logger.info("VP model only");
-			for (final Triple t : triples) {
-				final String tableName = Stats.getInstance().findTableName(t.getPredicate().toString());
-				final Node newNode = new VpNode(new TriplePattern(t, prefixes), tableName);
-				nodesQueue.add(newNode);
-			}
+			createVpNodes(triples, nodesQueue);
 		}
 		return nodesQueue;
 	}
 
+	/**
+	 * Given a <code>JoinedTriplesGroup</code>, creates its JWPT node, or VP nodes for each
+	 * element.
+	 *
+	 * @param joinedGroup
+	 *            <code>JoinedTriplesGroup</code> object with the lists of WPT and IWPT
+	 *            groups to be added to a single JWPT node, or to individual VP nodes
+	 * @param nodesQueue
+	 *            PriorityQueue where created nodes are added to
+	 */
+	private void createJwptVpNodes(final JoinedTriplesGroup joinedGroup, final PriorityQueue<Node> nodesQueue) {
+		if (joinedGroup.getIwptGroup().size() + joinedGroup.getWptGroup().size() >= minimumGroupSize) {
+			nodesQueue.add(new JptNode(joinedGroup, prefixes));
+		} else {
+			createVpNodes(new ArrayList<>(joinedGroup.getWptGroup()), nodesQueue);
```

Comment:
If you add the smallest set then the complexity of both will be the same. I wasn't so much concern about the complexity, but the readability. You would have just invoke createVpNodes on the joined set. Anyway, this is not such a big issue,if you prefer it your way, feel free to leave it like it is now.

Reply:
Wouldnt that make the code less readable?

```
HashSet<Triple> union = new HashSet<>();
if (joinedGroup.getWPTGroup().size() > joinedGroup.getIWTGroup().size() {
	union.addAll(joinedGroup.getWPTGroup());
	union.addAll(joinedGroup.getIWTGroup());
} else {
	union.addAll(joinedGroup.getIWTGroup());
	union.addAll(joinedGroup.getWPTGroup());
}
createVpNodes(new ArrayList<>(union), nodesQueue);
```

Or what exactly are you proposing?

Reply:
I didn't think of the need to have a new union hashset. I agree that way with removing triples is better. No need of change.

Reply:
Yes, I checked the rest of the code and the object is deleted after the VP nodes are created. I didn't want to addAll direct to the list, because then we would have wrong elements in the list, but that is not problem if we make sure we do not use it after that.
Anyways, removeAll is slightly faster, so there's no need to change that.